### ISSUE TITLE
`roc glue` with rust spec will not call exposed effects correctly 

### ISSUE THREAD

**Issue by JRMurr** [roc-lang/roc#6288]
This [commit of my repo](https://github.com/JRMurr/roc-pbt/tree/54111f798846f19a688d0c6227ba53645e3c8fdb) has the issue.

My platform exposes some functions but if i call any of them from roc the program would just end.

If i don't use the generated glue logic in [roc_app](https://github.com/JRMurr/roc-pbt/blob/54111f798846f19a688d0c6227ba53645e3c8fdb/platform/crates/glue/roc_app/src/x86_64.rs) with this main function
```rust
#[no_mangle]
pub extern "C" fn rust_main() -> i32 {
    println!("ENTERTING ROC");
    let res = mainForHost();
    println!("LEAVING ROC\nRES:({res:?})");
    0
}
```

and instead use the main function from basic cli 

```rust
#[no_mangle]
pub extern "C" fn rust_main() -> i32 {
    let size = unsafe { roc_main_size() } as usize;
    let layout = Layout::array::<u8>(size).unwrap();

    unsafe {
        let buffer = std::alloc::alloc(layout);

        roc_main(buffer);

        call_the_closure(buffer);

        std::alloc::dealloc(buffer, layout);
    }

    0
}

extern "C" {
    #[link_name = "roc__mainForHost_1_exposed_generic"]
    pub fn roc_main(output: *mut u8);

    #[link_name = "roc__mainForHost_1_exposed_size"]
    pub fn roc_main_size() -> i64;

    #[link_name = "roc__mainForHost_0_caller"]
    fn call_Fx(flags: *const u8, closure_data: *const u8, output: *mut u8);

    #[allow(dead_code)]
    #[link_name = "roc__mainForHost_0_size"]
    fn size_Fx() -> i64;

    #[link_name = "roc__mainForHost_0_result_size"]
    fn size_Fx_result() -> i64;
}

pub unsafe fn call_the_closure(closure_data_ptr: *const u8) -> u8 {
    let size = size_Fx_result() as usize;
    let layout = Layout::array::<u8>(size).unwrap();
    let buffer = std::alloc::alloc(layout);

    call_Fx(
        // This flags pointer will never get dereferenced
        MaybeUninit::uninit().as_ptr(),
        closure_data_ptr,
        buffer,
    );

    std::alloc::dealloc(buffer, layout);

    // TODO return the u8 exit code returned by the Fx closure
    0
}
```

the effects work as expected

**Comment by JRMurr** [comment]
Its also very possible i mis-used the generated glue code in my main (the first one) so I can close if so

**Comment by JRMurr** [comment]
some extra context from zulip https://roc.zulipchat.com/#narrow/stream/302903-Writing-a-platform/topic/generated.20rust.20glue.20seems.20different.20from.20whats.20in.20basic-cli/near/408304313

TLDR: the issue seems to be that the glue code is treating effect as a value but it really should treat it similar to a function.
The basic-webserver works around this by having a [main-glue](https://github.com/roc-lang/basic-webserver/blob/main/platform/main-glue.roc) with a different type signature
